exponentially weighted Bessel function test did not work under 32 bit Linux

jschnetm opened the pull request:
Hi Luigi - this is a patch for the test which had previously thrown errors under 32 bit ubuntu 15.04. I think that I has not broken the test under 64 bit, but it is probably better if you would retest that,

pcaspers commented:
Hi Johannes, I fear that I am guilty since I added these weighted variants to Klaus' implementation. However I do not get the actual problem. Can you explain that in a few sentences maybe ? Thank you, Peter

jschnetm commented:
Hi Peter,

it seems to be a rounding problem in the test. I have reformulated the test a 
bit (pull request #266) and now the tests run through. Apart from some eye 
candy and added "consts" the core of the patch is to extract the exponential 
weight from the sum as:
                M_PI_2 * (modifiedBesselFunction_i(-nu,x) -
                          modifiedBesselFunction_i(nu,x)) * exp(-x) / 
std::sin(M_PI*nu);

Secondly the test tolerance will now not go below QL_EPSILON, which is needed 
for the tests of the very small results.

Kind regards,
Johannes

On Friday, June 12, 2015 07:44:02 AM Peter Caspers wrote:
> Hi Johannes, I fear that I am guilty since I added these weighted variants
> to Klaus' implementation. However I do not get the actual problem. Can you
> explain that in a few sentences maybe ? Thank you, Peter
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/lballabio/quantlib/pull/266#issuecomment-111515800

pcaspers commented:
yes, I get the part with factoring out the exp(-x) term for the expected result. Why do we need to floor the tolerance for the k-case at QL_EPSILON though ?

jschnetm commented:
On Friday, June 12, 2015 09:51:29 AM Peter Caspers wrote:
> yes, I get the part with factoring out the exp(-x) term for the expected
> result. Why do we need to floor the tolerance for the k-case at QL_EPSILON
> though ?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/lballabio/quantlib/pull/266#issuecomment-111554589

Otherwise the tests fail if the result is small. An example (I have included 
the tol, which results from the other formula):

functions.cpp(267): error in 
"QuantLib::detail::quantlib_test_case(&FunctionsTest::testWeightedModifiedBesselFunctions)": 
failed to verify exponentially weightedmodified Bessel function of second kind
 order      : -4.5
 argument   : 11.1
 calculated  : 2.01882e-10
 expected   : 2.01882e-10
 tol        : 2.96624e-18

pcaspers commented:
ok. Thank you, Johannes !

On 12 June 2015 at 20:45, Johannes Göttker-Schnetmann <
notifications@github.com> wrote:

> On Friday, June 12, 2015 09:51:29 AM Peter Caspers wrote:
> > yes, I get the part with factoring out the exp(-x) term for the expected
> > result. Why do we need to floor the tolerance for the k-case at
> QL_EPSILON
> > though ?
> >
> > ---
> > Reply to this email directly or view it on GitHub:
> > https://github.com/lballabio/quantlib/pull/266#issuecomment-111554589
>
> Otherwise the tests fail if the result is small. An example (I have
> included
> the tol, which results from the other formula):
>
> functions.cpp(267): error in
> "QuantLib::detail::quantlib_test_case(&FunctionsTest::testWeightedModifiedBesselFunctions)":
>
> failed to verify exponentially weightedmodified Bessel function of second
> kind
> order : -4.5
> argument : 11.1
> calculated : 2.01882e-10
> expected : 2.01882e-10
> tol : 2.96624e-18
>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/lballabio/quantlib/pull/266#issuecomment-111586560>.
>
